File: Projects/Autocal/attic/autocal-20100821/src/loader.cpp
#include "loader.h"
#include <QFile>
#include <QRegExp>
#include <QStringList>
#include <QTextStream>

Loader::Loader(QObject *parent):
	QObject(parent), _lineNumber(-1)
{
}

int Loader::lineNumber()
{
	return _lineNumber;
}

void Loader::saveToFile(Schedule *schedule, const QString &filename)
{
	QFile file(filename);
	file.open(QIODevice::WriteOnly | QIODevice::Text);

	QTextStream ts(&file);
	ts << _saveHelper(schedule);
	file.close();
}

QString Loader::_saveHelper(Schedule *schedule)
{
	QString d;
	Task *task;

	for (int i = 0; i < schedule->taskList().length(); i++)
	{
		task = schedule->taskList().at(i);
		d.append("{\n");
		d.append(QString("\tbegin = %1\n").arg(task->beginTime()));
		d.append(QString("\tduration = %1\n").arg(task->durationTime()));
		d.append(QString("\tdeadline = %1\n").arg(task->deadlineTime()));
		d.append(QString("\tdescription = %1\n").arg(task->description()));
		d.append(QString("\tfixed = %1\n").arg(
			(task->fixed() == true ? "yes" : "no")));
		d.append("}\n\n");
	}

	return d;
}

Schedule *Loader::loadFromFile(const QString &filename)
{
	QFile file(filename);

	if (file.exists())
		file.open(QIODevice::ReadOnly | QIODevice::Text);
	else
		return 0;

	QString data(file.readAll());
	file.close();
	return _loadHelper(data);
}

Schedule *Loader::_loadHelper(const QString &data)
{
	QStringList lines = data.split('\n');
	QString line, description, ppName, ppValue;
	uint beginTime = 0, durationTime = 0, deadlineTime = 0;
	bool fixed = false;
	QRegExp rx("^([a-zA-Z]+)[\\s]*=[\\s]*(\\w[\\w\\s]*)$", Qt::CaseInsensitive);
	Task *task;
	Schedule *schedule = new Schedule();

	for (int i = 0; i < lines.length(); i++)
	{
		line = lines.at(i).trimmed();

		if (line.startsWith('#') || line.isEmpty())
			continue;

		if (line.startsWith('{'))
		{
			beginTime = 0;
			durationTime = 0;
			deadlineTime = 0;
			description = "";
			fixed = false;

			continue;
		}

		if (line.startsWith('}'))
		{
			task = new Task(schedule);
			task->setBeginTime(beginTime);
			task->setDurationTime(durationTime);
			task->setDeadlineTime(deadlineTime);
			task->setDescription(description);
			task->setFixed(fixed);
			schedule->addTask(task);

			continue;
		}

		if (rx.indexIn(line) > -1)
		{
			ppName = rx.cap(1);
			ppValue = rx.cap(2);

			if (ppName == "begin")
				beginTime = ppValue.toUInt();
			else if (ppName == "duration")
				durationTime = ppValue.toUInt();
			else if (ppName == "deadline")
				deadlineTime = ppValue.toUInt();
			else if (ppName == "description")
				description = ppValue;
			else if (ppName == "fixed")
				fixed = (ppValue == "yes" ? true : false);
			else
				goto leave;

			continue;
		}

		leave:
		delete schedule;
		_lineNumber = i;
		return 0;
	}

	return schedule;
}

File: Projects/Autocal/attic/autocal-py/libautocal/loader.py
import re
from config import _error
from schedule import Schedule, Task

def save(sched):
	res = ''
	task = '{\n\tdescription = %s\n\tdeadline = %s\n\tbegin = %s\n\tduration = %s\n\tfixed = %s\n}\n'
	for t in sched.tasks:
		res += (task % (str(t.id), t.description, str(t.deadline), str(t.begin), str(t.duration), '0'))
	for t in sched.fixedtasks:
		res += (task % (str(t.id), t.description, str(t.deadline), str(t.begin), str(t.duration), '1'))
	return res

def load(data):
	id = None
	description = None
	deadline = None
	begin = None
	duration = None
	fixed = None

	_sched = Schedule()

	for _lineno, _line in enumerate(data.split('\n')):
		line = _line.strip()
		if not line.startswith('#') and line != '':
			regexp_res = re.match(r'(?P<task_id>[0-9]+)[ \t]*{', line)
			if regexp_res:
				id = int(regexp_res.group('task_id'))
				continue

			regexp_res = re.match(
				r'(?P<prop_name>[a-zA-Z]+)[ \t]*=[ \t]*(?P<prop_val>.*)', line)
			if regexp_res:
				if regexp_res.group('prop_name') == 'description':
					description = str(regexp_res.group('prop_val'))
				elif regexp_res.group('prop_name') == 'deadline':
					deadline = int(regexp_res.group('prop_val'))
				elif regexp_res.group('prop_name') == 'begin':
					begin = int(regexp_res.group('prop_val'))
				elif regexp_res.group('prop_name') == 'duration':
					duration = int(regexp_res.group('prop_val'))
				elif regexp_res.group('prop_name') == 'fixed':
					fixed = int(regexp_res.group('prop_val'))
				continue

			if line == '}':
				t = Task(id, begin, duration, deadline, description, fixed)
				if fixed == 0:
					_sched.tasks += [t]
				else:
					_sched.fixedtasks += [t]
				continue

			_error('Error at line ' + str(_lineno))
	return _sched
